File: gui/src/main/java/com/woohyman/gui/ui/NesGalleryActivity.java
package com.woohyman.gui.ui;

import android.annotation.TargetApi;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;

import com.woohyman.gui.base.EmulatorActivity;
import com.woohyman.gui.base.OpenGLTestActivity;
import com.woohyman.gui.ui.gamegallery.GalleryActivity;

import java.util.HashSet;
import java.util.Set;

import com.woohyman.keyboard.emulator.Emulator;
import com.woohyman.keyboard.utils.EmuUtils;
import com.woohyman.keyboard.utils.NLog;
import com.woohyman.keyboard.utils.PreferenceUtil;

public class NesGalleryActivity extends GalleryActivity {

    private static final int REQUEST_CHECK_OPENGL = 200;

    @Override
    public Emulator getEmulatorInstance() {
        return NesEmulator.getInstance();
    }

    @Override
    public Class<? extends EmulatorActivity> getEmulatorActivityClass() {
        return NesEmulatorActivity.class;
    }

    @Override
    protected Set<String> getRomExtensions() {
        HashSet<String> set = new HashSet<>();
        set.add("nes");
        set.add("fds");
        return set;
    }


    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (PreferenceUtil.getFragmentShader(this) == -1
                && EmuUtils.checkGL20Support(this)) {
            Intent intent = new Intent(this, OpenGLTestActivity.class);
            startActivityForResult(intent, REQUEST_CHECK_OPENGL);
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_CHECK_OPENGL) {
            NLog.e("opengl", "opengl: " + resultCode);
            PreferenceUtil.setFragmentShader(this, resultCode);
        }
    }
}

File: gui/xml/src/main/java/com/woohyman/xml/ui/NesGalleryActivity.kt
package com.woohyman.xml.ui

import android.content.Intent
import android.os.Bundle
import com.woohyman.xml.base.EmulatorActivity
import com.woohyman.xml.base.OpenGLTestActivity
import com.woohyman.xml.ui.gamegallery.GalleryActivity
import com.woohyman.keyboard.emulator.Emulator
import com.woohyman.keyboard.emulator.NesEmulator
import com.woohyman.keyboard.utils.EmuUtils.checkGL20Support
import com.woohyman.keyboard.utils.NLog.e
import com.woohyman.keyboard.utils.PreferenceUtil.getFragmentShader
import com.woohyman.keyboard.utils.PreferenceUtil.setFragmentShader
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

@AndroidEntryPoint
class NesGalleryActivity : GalleryActivity() {

    @Inject
    lateinit var nesEmulator: NesEmulator

    override val emulatorActivityClass: Class<out EmulatorActivity?>?
        get() = NesEmulatorActivity::class.java

    override val romExtensions: Set<String>?
        get() = kotlin.run {
            val set = HashSet<String>()
            set.add("nes")
            set.add("fds")
            return set
        }

    override val emulatorInstance: Emulator?
        get() = nesEmulator

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (getFragmentShader(this) == -1
            && checkGL20Support(this)
        ) {
            val intent = Intent(this, OpenGLTestActivity::class.java)
            startActivityForResult(intent, REQUEST_CHECK_OPENGL)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_CHECK_OPENGL) {
            e("opengl", "opengl: $resultCode")
            setFragmentShader(this, resultCode)
        }
    }

    companion object {
        private const val REQUEST_CHECK_OPENGL = 200
    }
}